qiskit_experiments/library/driven_freq_tuning/coefficient_utils.py

```diff
@@ -0,0 +1,234 @@
+            valid_amp = float(valid_amps[before_inflection].real)
+        else:
+            valid_amp = float(valid_amps[0].real)
+        amplitudes[idx] = min(valid_amp, 1.0)
+    return amplitudes
+
+
+def convert_amp_to_freq(
+    amps: np.ndarray,
+    coeffs: StarkCoefficients,
+) -> np.ndarray:
+    """A helper function to convert Stark amplitude to frequency shift.
+
+    Args:
+        amps: Amplitude values to convert into frequency shift.
+        coeffs: Calibrated Stark coefficients.
+
+    Returns:
+          Calculated frequency shift at given Stark amplitude.
+    """
+    pos_fit = np.poly1d([*coeffs.positive_coeffs(), coeffs.offset])
+    neg_fit = np.poly1d([*coeffs.negative_coeffs(), coeffs.offset])
+
+    return np.where(amps > 0, pos_fit(amps), neg_fit(amps))
+
+
+def find_min_max_frequency(
+    min_amp: float,
+    max_amp: float,
+    coeffs: StarkCoefficients,
+) -> tuple[float, float]:
+    """A helper function to estimate maximum frequency shift within given amplitude budget.
+
+    Args:
+        min_amp: Minimum Stark amplitude.
+        max_amp: Maximum Stark amplitude.
+        coeffs: Calibrated Stark coefficients.
+
+    Returns:
+        Minimum and maximum frequency shift available within the amplitude range.
+    """
+    trim_amps = []
+    for amp in [min_amp, max_amp]:
+        if amp > 0:
+            fit = coeffs.positive_coeffs()
+        else:
+            fit = coeffs.negative_coeffs()
+        # Solve for inflection points by computing the point where derivative becomes zero.
+        solutions = np.roots([deriv * coeff for deriv, coeff in zip((3, 2, 1), fit)])
+        inflection_points = solutions[(solutions.imag == 0) & (np.sign(solutions) == np.sign(amp))]
+        if len(inflection_points) > 0:
+            # When multiple inflection points are found, use the most outer one.
+            # There could be a small inflection point around amp=0,
+            # when the first order term is significant.
+            amp = min([amp, max(inflection_points, key=abs)], key=abs)
+        trim_amps.append(amp)
+    return tuple(convert_amp_to_freq(np.asarray(trim_amps), coeffs))
+
+
+def retrieve_coefficients_from_service(
+    service: IBMExperimentService,
+    backend_name: str,
+    qubit: int,
+) -> StarkCoefficients:
+    """Retrieve StarkCoefficients object from experiment service.
+
+    Args:
+        service: IBM Experiment service instance interfacing with result database.
+        backend_name: Name of target backend.
+        qubit: Index of qubit.
+
+    Returns:
+        StarkCoefficients object.
+
+    Raises:
+        RuntimeError: When stark_coefficients entry doesn't exist in the service.
+    """
+    try:
+        retrieved = service.analysis_results(
+            device_components=[f"Q{qubit}"],
+            result_type="stark_coefficients",
+            backend_name=backend_name,
+            sort_by=["creation_datetime:desc"],
+        )
```

Comment:
Something is off here for me. `IBMExperimentService` does not know about the `StarkCoefficients` class so it should be able to return an instance of it from `analysis_results()` here like the type hint suggests. I tried adding `json_decoder=ExperimentDecoder`, but that completely solve the issue. When I do that, I get a `retrieved[0].result_data` like:

```python
{'value': '(StarkCoefficients)',
 '_chisq': 8.026329774009517,
 '_extra': {'run_time': None, 'experiment': 'StarkRamseyXYAmpScan'},
 '_value': StarkCoefficients(pos_coef_o1=5748041.745393166, pos_coef_o2=17930671.766640767, pos_coef_o3=7365555.846512856, neg_coef_o1=-581564.1840185896, neg_coef_o2=-27118377.31311793, neg_coef_o3=3073801.6774436794, offset=-658782.8378606051),
 '_source': {'class': 'qiskit_experiments.framework.analysis_result.AnalysisResult',
  'data_version': 1,
  'qiskit_version': {'qiskit': '0.45.1', 'qiskit-experiments': '0.6.0'}}}
```

So it does have the `StarkCoefficients` instance in `_value` but it is a dict and `'value'` is just a string `(StarkCoefficients)`.

If I don't set the decoder I get

```python
{'value': '(StarkCoefficients)',
 '_chisq': 8.026329774009517,
 '_extra': {'run_time': None, 'experiment': 'StarkRamseyXYAmpScan'},
 '_value': {'__type__': 'object',
  '__value__': {'class': {'__type__': 'type',
    '__value__': {'name': 'StarkCoefficients',
     'module': 'qiskit_experiments.library.driven_freq_tuning.coefficient_utils',
     'version': '0.6.0.dev0+da444b6'}},
   'version': '0.6.0.dev0+da444b6',
   'settings': {'offset': -658782.8378606051,
    'neg_coef_o1': -581564.1840185896,
    'neg_coef_o2': -27118377.31311793,
    'neg_coef_o3': 3073801.6774436794,
    'pos_coef_o1': 5748041.745393166,
    'pos_coef_o2': 17930671.766640767,
    'pos_coef_o3': 7365555.846512856}}},
 '_source': {'class': 'qiskit_experiments.framework.analysis_result.AnalysisResult',
  'data_version': 1,
  'qiskit_version': {'qiskit': '0.45.1', 'qiskit-experiments': '0.6.0'}}}
```

Reply:
Oh thanks for catching this. I was thinking the fake service reproduces the behavior of the IBM service, but details are different in many places. For now I fixed the problem in [357b13d](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1236/commits/357b13d3c82b39c342f1a5237e0699c34fbbd713) and added minimum example code I tested with a real server in [12ee094](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1236/commits/12ee094fad890d4dd77fc596670e90e87409e443) as a tutorial.

Reply:
fyi https://github.com/Qiskit-Extensions/qiskit-experiments/issues/1383